pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -224,6 +151,269 @@ func (a *AmbientIndex) matchesScope(scope model.WaypointScope, w *model.Workload
+			// Configured for override
+			wl.Protocol = workloadapi.Protocol_HTTP
+		}
+
+		// Otherwise supports tunnel directly
+		if model.SupportsTunnel(pod.Labels, model.TunnelHTTP) {
+			wl.Protocol = workloadapi.Protocol_HTTP
+			wl.NativeHbone = true
+		}
+		wl.AuthorizationPolicies = a.controller.selectorAuthorizationPolicies(namespace, labels)
+	} else if entry != nil {
+		labels = entry.Labels
+		wl.Name = configEntry.Name
+		wl.ServiceAccount = entry.ServiceAccount
+		if td := spiffe.GetTrustDomain(); td != "cluster.local" {
+			wl.TrustDomain = td
+		}
+
+		// TODO: support WorkloadGroup
+		// wl.WorkloadName, wl.WorkloadType = workloadNameAndType(pod)
+		wl.CanonicalName, wl.CanonicalRevision = kubelabels.CanonicalService(labels, wl.WorkloadName)
+
+		if configEntry.Annotations[constants.AmbientRedirection] == constants.AmbientRedirectionEnabled {
+			// Configured for override
+			wl.Protocol = workloadapi.Protocol_HTTP
+		}
+
+		// Otherwise supports tunnel directly
+		if model.SupportsTunnel(labels, model.TunnelHTTP) {
+			wl.Protocol = workloadapi.Protocol_HTTP
+			wl.NativeHbone = true
+		}
+		wl.AuthorizationPolicies = a.controller.selectorAuthorizationPolicies(namespace, labels)
+	}
+
+	// Setup waypoints.
+	a.buildWaypointAddresses(labels, wl)
+
+	return &model.WorkloadInfo{Workload: wl, Labels: labels}
+}
+
+func (a *AmbientIndex) buildWaypointAddresses(labels map[string]string, wl *workloadapi.Workload) {
+	if isWaypoint(labels) {
+		// Waypoints do not have waypoints themselves, so filter them.
+		return
+	}
+	var waypoints []*v1.Pod
+	// First check for a waypoint for our SA explicit
+	// TODO: this is not robust against temporary waypoint downtime. We also need the users intent (Gateway).
+	waypoints = a.waypointsIndex.Lookup(model.WaypointScope{Namespace: wl.Namespace, ServiceAccount: wl.ServiceAccount})
+	if len(waypoints) == 0 {
+		// if there are none, check namespace wide waypoints
+		waypoints = a.waypointsIndex.Lookup(model.WaypointScope{Namespace: wl.Namespace})
+	}
+
+	// If we have a waypoint, configure it
+	if len(waypoints) > 0 {
+		ips := make([][]byte, 0, len(waypoints))
+		for _, r := range waypoints {
+			if ip, err := netip.ParseAddr(r.Status.PodIP); err == nil {
+				ips = append(ips, ip.AsSlice())
+			}
+		}
+		// Keep things stable
+		slices.SortFunc(ips, func(a, b []byte) bool {
+			return string(a) < string(b)
+		})
+		wl.WaypointAddresses = ips
+	}
+}
+
+// buildVIPsFromWorkloadEntry adds vips to `vips` based on WorkloadEntry
+// Note: this mutates `vips`!
+func (a *AmbientIndex) buildVIPsFromWorkloadEntry(entry *networking.WorkloadEntry, namespace string, vips map[string]*workloadapi.PortList) {
+	if entry != nil {
+		// TODO: can we derive this from ServiceInstances from the various existing stores instead of redoing everything here?
+		// find the workload entry's service by label selector
+		// rather than scanning through our internal map of model.services, get the services via the k8s apis
```

Comment:
i think we may want to do this in the end once we support service entries though.. especially to avoid redoing auto dns allocation. i think we may want to hook into existing stores to avoid recomputation